pkg/sentry/platform/ring0/pagetables/pagetables_amd64.go

```diff
@@ -50,5 +50,12 @@ func (p *PageTables) Init(allocator Allocator) {
 	p.rootPhysical = p.Allocator.PhysicalFor(p.root)
 }
 
+// cloneUpperHalf clone the upper half from the upper half page tables
+//
+//go:nosplit
+func (p *PageTables) cloneUpperHalf() {
+	copy(p.root[entriesPerPage/2:entriesPerPage], p.upperHalfPageTables.root[entriesPerPage/2:entriesPerPage])
```

Comment:
If you allow a dynamic upperLimit, you should check here that it is outside a canonical range, and that it can be appropriate split at the pgd level. I'm sure there are a similar checks that might apply to ARM.

You should document these constraints on the NewWithUpper function as Preconditions.

Reply:
If a dynamic upperLimit is allowed, there will be complicated to calculate the index which depends on which real halfside the upperLimit belongs to and depends on the canonical range and the page level (4 levels or 5 levels of pagetables). But gvisor has notcode to handle 5 levels of pagetables yet, so I'd like to wait until 5 levels of pagetables has been supported #4696 would I allow a dynamic upperLimit.